java/arcs/sdk/android/storage/AndroidStorageServiceEndpointManager.kt

```diff
@@ -1,20 +1,121 @@
-
-) = DirectStorageEndpointManager(
-  StoreManager(
-    ServiceStoreFactory(context, coroutineContext, connectionFactory)
-  )
-)
+class AndroidStorageServiceEndpointManager(
+  private val coroutineScope: CoroutineScope,
+  private val connectionFactory: ConnectionFactory
+) : StorageEndpointManager {
+
+  override suspend fun <Data : CrdtData, Op : CrdtOperationAtTime, T> get(
+    storeOptions: StoreOptions,
+    callback: ProxyCallback<Data, Op, T>
+  ): StorageEndpoint<Data, Op, T> {
+    val connection = connectionFactory(storeOptions, storeOptions.type.toParcelableType())
+    val service = IStorageService.Stub.asInterface(connection.connectAsync().await())
+    val channelId = suspendForRegistrationCallback { resultCallback ->
+      service.registerCallback(
+        StorageServiceProxyCallback(coroutineScope, callback),
+        resultCallback
+      )
+    }
+    return AndroidStorageEndpoint(
+      channelId, service, connection
+    )
+  }
+
+  override suspend fun reset() {}
+}
+
+@ExperimentalCoroutinesApi
+class AndroidStorageEndpoint<Data : CrdtData, Op : CrdtOperationAtTime, T>(
+  private val channelId: Int,
+  private val service: IStorageService,
+  private val connection: StorageServiceConnection
+) : StorageEndpoint<Data, Op, T> {
+  private val outgoingMessagesCount = CounterFlow(0)
+
+  val log = TaggedLog { "AndroidStorageEndpoint" }
+
+  override suspend fun idle() {
+    log.debug { "Waiting for service store to be idle" }
+    outgoingMessagesCount.flow.first { it == 0 }
+    suspendForResultCallback { resultCallback ->
+      service.idle(TIMEOUT_IDLE_WAIT_MILLIS, resultCallback)
+    }
+    log.debug { "Endpoint is idle" }
+  }
+
+  override suspend fun onProxyMessage(message: ProxyMessage<Data, Op, T>) {
+    outgoingMessagesCount.increment()
+    try {
+      suspendForResultCallback { resultCallback ->
+        service.sendProxyMessage(
+          message.withId(channelId).toProto().toByteArray(), resultCallback
+        )
+      }
+    } catch (e: CrdtException) {
+      // Just return false if the message couldn't be applied.
+      log.debug(e) { "CrdtException occurred in onProxyMessage" }
+    } finally {
+      outgoingMessagesCount.decrement()
+    }
+  }
+
+  override suspend fun close() {
+    suspendForResultCallback { resultCallback ->
+      service.unregisterCallback(channelId, resultCallback)
+    }
+    connection.disconnect()
+  }
+
+  companion object {
+    private const val TIMEOUT_IDLE_WAIT_MILLIS = 10000L
+  }
+}
+
+class StorageServiceProxyCallback<Data : CrdtData, Op : CrdtOperation, T>(
+  private val scope: CoroutineScope,
+  private val callback: ProxyCallback<Data, Op, T>
+) : IStorageServiceCallback.Stub() {
+  override fun onProxyMessage(proxyMessage: ByteArray) {
+    scope.launch {
+      @Suppress("UNCHECKED_CAST")
+      callback(
+        proxyMessage.decodeProxyMessage()
+          as ProxyMessage<Data, Op, T>
+      )
+    }
+  }
```

Comment:
To ensure that the launches don't happen out-of-order, at minimum we should use `start = CoroutineStart.UNDISPATCHED`... maybe better (?) would be to feed the messages to a channel that gets drained in-order using calls to the callback.

Reply:
This should eventually be a queue (I think it's mentioned in my queues survey doc).